modules/effects/src/main/scala/dev/profunktor/redis4cats/transactions.scala

```diff
@@ -39,25 +44,56 @@ object transactions {
       * It should not be used to run other computations, only Redis commands. Fail to do so
       * may end in unexpected results such as a dead lock.
       */
-    def run(commands: F[Any]*): F[Unit] =
-      Ref.of[F, List[Fiber[F, Any]]](List.empty).flatMap { fibers =>
-        val tx =
-          Resource.makeCase(cmd.multi) {
-            case (_, ExitCase.Completed) =>
-              cmd.exec *> F.info("Transaction completed")
-            case (_, ExitCase.Error(e)) =>
-              cmd.discard *> F.error(s"Transaction failed: ${e.getMessage}")
-            case (_, ExitCase.Canceled) =>
-              cmd.discard *> F.error("Transaction canceled")
+    def exec[T <: HList, R <: HList](commands: T)(implicit w: Witness.Aux[T, R]): F[R] =
+      Deferred[F, Either[Throwable, w.R]].flatMap { promise =>
+        // Forks every command in order
+        def runner[H <: HList, G <: HList](ys: H, res: G): F[Any] =
+          ys match {
+            case HNil                           => F.pure(res)
+            case HCons((h: F[_] @unchecked), t) => h.start.flatMap(fb => runner(t, fb :: res))
+          }
+
+        // Joins or cancel fibers correspondent to previous executed commands
+        def joinOrCancel[H <: HList, G <: HList](ys: H, res: G)(isJoin: Boolean): F[Any] =
+          ys match {
+            case HNil => F.pure(res)
+            case HCons((h: Fiber[F, Any] @unchecked), t) if isJoin =>
+              F.info(">>> Got fiber") >>
+                  h.join.flatMap(x => F.info(s">>> Content: $x") >> joinOrCancel(t, x :: res)(isJoin))
+            case HCons((h: Fiber[F, Any] @unchecked), t) =>
+              h.cancel.flatMap(x => joinOrCancel(t, x :: res)(isJoin))
+            case HCons(h, t) =>
+              F.error(s">>> Unexpected cons: ${h.toString}") >> joinOrCancel(t, res)(isJoin)
           }
 
-        val cancelFibers =
-          fibers.get.flatMap(_.traverse(_.cancel).void)
+        def cancelFibers(fibs: HList, err: Throwable = TransactionAborted): F[Unit] =
+          joinOrCancel(fibs.reverse, HNil)(false).void >> promise.complete(err.asLeft)
 
-        F.info("Transaction started") *>
-          tx.use(_ => commands.toList.traverse(_.start).flatMap(fibers.set))
-            .guarantee(cancelFibers)
+        val tx =
+          Resource.makeCase(cmd.multi >> runner(commands, HNil)) {
+            case ((fibs: HList), ExitCase.Completed) =>
+              for {
+                _ <- F.info("Transaction completed")
+                _ <- cmd.exec.handleErrorWith(e => cancelFibers(fibs.reverse, e) >> F.raiseError(e))
+                tr <- joinOrCancel(fibs.reverse, HNil)(true)
+                // Casting here is fine since we have a `Witness` that proves this true
```

Comment:
Maybe the witness type could have a conversion function in it so we could avoid the cast?

Reply:
Not sure how that can be done, to be honest. `Witness` only operates at the type level, no values are ever managed.

Reply:
yeah, we'd need to change it. The whole thing looks remarkably like traverse :P